### PR TITLE
Async API

### PR DESCRIPTION
**Pull request by ankit-iitb**

### CODE REVIEW

**Comment by ankit-iitb** [comment]
This PR resolves https://github.com/vmware/node-replication/issues/20

File: benches/hashmap_comparisons.rs [hunk 1]
```diff
@@ -140,6 +161,14 @@ where
         unreachable!("All opertations must be read ops")
     }
 
+    async fn async_exec(
+        &self,
+        _op: <Self::D as Dispatch>::WriteOperation,
+        _idx: ReplicaToken,
+    ) -> Pin<Box<dyn futures::Future<Output = <Self::D as Dispatch>::Response> + 'life0>> {
+        unreachable!("All opertations must be read ops")
```
Comment (gz) [inline benches/hashmap_comparisons.rs:169]: typo operations

File: benches/hashmap_comparisons.rs [hunk 2]
```diff
@@ -155,6 +184,14 @@ where
     ) -> <Self::D as Dispatch>::Response {
         self.data_structure.dispatch(op)
     }
+
+    async fn async_exec_ro(
+        &self,
+        _op: <Self::D as Dispatch>::ReadOperation,
+        _idx: ReplicaToken,
+    ) -> Pin<Box<dyn futures::Future<Output = <Self::D as Dispatch>::Response> + 'life0>> {
+        unreachable!("All opertations must be read ops")
```
Comment (gz) [inline benches/hashmap_comparisons.rs:193]: typo operations

File: benches/hashmap_comparisons.rs [hunk 3]
```diff
@@ -74,6 +78,14 @@ where
         unsafe { (&mut *self.data_structure.get()).dispatch_mut(op) }
     }
 
+    async fn async_exec(
+        &self,
+        _op: <Self::D as Dispatch>::WriteOperation,
+        _idx: ReplicaToken,
+    ) -> Pin<Box<dyn futures::Future<Output = <Self::D as Dispatch>::Response> + 'life0>> {
+        unreachable!("All opertations must be sync ops")
```
Comment (gz) [inline benches/hashmap_comparisons.rs:86]: typo

File: benches/mkbench.rs [hunk 4]
```diff
@@ -54,11 +58,16 @@ type BenchFn<R> = fn(
...
-        <<R as ReplicaTrait>::D as Dispatch>::ReadOperation,
-        <<R as ReplicaTrait>::D as Dispatch>::WriteOperation,
+    &Vec<
+        Operation<
+            <<R as ReplicaTrait>::D as Dispatch>::ReadOperation,
+            <<R as ReplicaTrait>::D as Dispatch>::WriteOperation,
+        >,
     >,
-    usize,
+    usize, // Operating vector length
```
Comment (gz) [inline benches/mkbench.rs:67]: Not 100% sure but I think we can name parameters like this: type myfun = fn(a: usize, b: usize),

we should probably give names now that the function is so huge?

File: benches/mkbench.rs [hunk 5]
```diff
@@ -67,13 +76,19 @@ type BenchFn<R> = fn(
     ReplicaToken,
     &Vec<Arc<Log<'static, <<R as ReplicaTrait>::D as Dispatch>::WriteOperation>>>,
     &Arc<R>,
-    &Operation<
-        <<R as ReplicaTrait>::D as Dispatch>::ReadOperation,
-        <<R as ReplicaTrait>::D as Dispatch>::WriteOperation,
+    &Vec<
```
Comment (gz) [inline benches/mkbench.rs:79]: same here named parameters might be good

File: benches/mkbench.rs [hunk 6]
```diff
@@ -153,6 +192,15 @@ impl<'a, T: Dispatch + Sync + Default> ReplicaTrait for Replica<'a, T> {
...
         self.execute(op, idx)
     }
+
+    #[cfg(feature = "nr")]
+    async fn async_exec_ro(
+        &self,
+        op: <Self::D as Dispatch>::ReadOperation,
+        idx: ReplicaToken,
+    ) -> Pin<Box<dyn futures::Future<Output = <Self::D as Dispatch>::Response> + 'life0>> {
+        Box::pin(self.async_execute(op, idx).await)
```
Comment (gz) [inline benches/mkbench.rs:202]: maybe explore passing in Option<Box::pin> as discussed to potentially avoid allocation?

File: nr/src/lib.rs [hunk 7]
```diff
@@ -111,12 +112,12 @@ pub trait Dispatch {
     /// A read-only operation. When executed against the data structure, an operation
     /// of this type must not mutate the data structure in anyway. Otherwise, the
     /// assumptions made by this library no longer hold.
-    type ReadOperation: Sized + Clone + PartialEq + Debug;
+    type ReadOperation: Sized + Clone + PartialEq + Debug + Send;
```
Comment (gz) [inline nr/src/lib.rs:115]: why is this necessary?

File: nr/src/lib.rs [hunk 8]
```diff
@@ -111,12 +112,12 @@ pub trait Dispatch {
...
 
     /// A write operation. When executed against the data structure, an operation of
     /// this type is allowed to mutate state. The library ensures that this is done so
     /// in a thread-safe manner.
-    type WriteOperation: Sized + Clone + PartialEq + Debug + Send;
+    type WriteOperation: Sized + Clone + PartialEq + Debug + Send + Sync;
```
Comment (gz) [inline nr/src/lib.rs:120]: why is this necessary?

**Review by gz — APPROVED** [review]

**Comment by gz** [comment]
Thanks great work :+1:
